Add more tests around using -f Dockerfile via the CLI

duglin opened the pull request:
Mainly to make sure we don't treat the -f value as relative to the
root of the build context, but instead it must be relative to the current
dir.

Signed-off-by: Doug Davis dug@us.ibm.com

Commit:
Add more tests around using -f Dockerfile via the CLI
Mainly to make sure we don't treat the -f value as relative to the
root of the build context, but instead it must be relative to the current
dir.

Signed-off-by: Doug Davis <dug@us.ibm.com>

LK4D4 commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -4526,85 +4527,85 @@ func TestBuildRenamedDockerfile(t *testing.T) {
 	}
 
-	out, _, err = dockerCmdInDir(t, "/tmp", "build", "-t", "test6", ctx.Dir)
+	out, _, err = dockerCmdInDir(t, "/tmp", "build", "-t", "test9", ctx.Dir)
```
Sorry I probably missed that last time, but you shouldn't use `/tmp`. Use `ioutil.TmpDir` and `defer os.RemoveAll`.

cpuguy83 replied:
`tmpDir, err := ioutil.TempDir(os.TempDir(), "my-test")`

LK4D4 replied:
I'm using `ioutil.TempDir("", "my-test")` because less symbols :)

cpuguy83 replied:
Oh yes, I fogot, if it's "" it automatically uses os.TempDir(), nice!

duglin replied:
Actually, just using os.TempDir() in place of "/tmp" is good enough. I'm not creating files in there, I'm just cd-ing to that dir so I can test running "docker build" from outside of the build context.  No need to create a new tempDir.

LK4D4 commented:
LGTM

tiborvass commented:
@duglin would you mind adding this too https://gist.github.com/tiborvass/8cfec4a2500195e94fad ?

It's a fix of one test I added.

duglin commented:
@tiborvass done - check to make sure I got all of it.

jfrazelle commented:
LGTM

tiborvass commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -4526,85 +4527,87 @@ func TestBuildRenamedDockerfile(t *testing.T) {
+		t.Fatalf("test10 should have used files/dFile2, output:%s", out)
+	}
+
+	out, _, err = dockerCmdInDir(t, tmpDir, "build", "-f", "Dockerfile", "-t", "test11", ctx.Dir)
```
@duglin I believe this is testing the same thing as the last few lines of this file. Can you reject or confirm?

tiborvass replied:
At `// Path to Dockerfile should be resolved relative to working directory, not relative to context.`

duglin replied:
yup I missed that, I'll remove this test.

tiborvass commented:
@duglin hey sorry to be annoying again, but I believe there are still duplicate tests. I'm also checking for `../outsideDockerfile`, maybe we should consolidate our tests into one?

duglin commented:
@tiborvass if you want... I'm not too concerned about too many tests :-)

duglin commented:
@tiborvass looking at the current list of testcases between those two funcs I see this:

FUNC: TestBuildRenamedDockerfile
files created:
```
context/Dockerfile
context/files/Dockerfile
context/files/dFile
context/dFile
context/files/dFile2
```
Tests: (dir run in, -f usage, context path provided, failure?)
```
A) from context     no -f                       .
B) from context     -f files/Dockerfile         .
C) from context     --file files/dFile          .
D) from context     --file dFile                .
E) from context     --file /etc/passwd          .         failure
F) from files       -f ../Dockerfile            ..
G) from files       -f /.../files/Dockerfile    ..
H) from files       -f ../Dockerfile            .         failure
J) from /tmp        no -f                       context
I) from /.../files  -f dFile2                   .
```

FUNC: TestBuildDockerfileOutsideContext
files created:
```
/tmp/context/Dockerfile  ENV X Y
/tmp/outsideDockerfile   ENV x y
/tmp/context/dockerfile1 -> /tmp/context/../outsideDockerfile
/tmp/context/dockerfile2 -> /tmp/outsideDockerfile
```

Tests:
```
J) from /tmp/context -f ../outsideDockerfile      .  failure
K) from /tmp/context -f /tmp/context/dockerfile1  .  failure
L) from /tmp/context -f /tmp/context/dockerfile2  .  failure
M) from /tmp         -f Dockerfile                /tmp/context/  failure
```

I think just J is a dup of E.  If you agree I can remove J.

tiborvass commented:
@duglin thanks so much!

Just to make sure we're talking about the same tests, I'm gonna remove ambiguity:
you put H) then J) then I), I'm gonna refer to those as H), J1) and I).
Later you put J) and K), I'm gonna refer to those as J2) and K).

A) seems to be the most normal build case, and it's covered by all tests using `buildImageFromContext`.
E) I would prefer not to use `/etc/hosts`, but rather a directory we create that's outside the context, like /tmp in TestBuildDockerfileOutsideContext. It's also not exactly a dup of J2).
G) not sure what that tests? `/.../` is weird
H) however is a dupe of J2)
I) I'm not sure what this tests for (again that weird `/.../`). Looks like D) to me

duglin commented:
@tiborvass - sorry about the typos - clearly I don't know my ABC's  :-)

A) yes its the normal case and therefore a dup of lots of tests, but I liked like to keep it because as I look at the full set of tests w.r.t. this feature I like having a complete stand-alone set in one spot. I'm sure there are plenty of dups like this if we looked across ALL of the testcases for all features.

I can change E) to use something other than /etc/passwd but to me its not the same as the other test because I wanted to use an abs path, not a relative one.  It shouldn't matter but I wanted to test it.

G) the /.../ syntax was just my way of saying that we specify a full abs path to the dockerfile, instead of a relative one. So, that might be a dup of others (like K and L) but the reason I added it was to make sure that the CLI didn't do something weird like pickup a Dockerfile from some place else in the tree above it.  Was thinking of how git will walk the tree to find files. I know we don't do this today so its a bit preemptive.

J2) yes we can probably remove that one as a dup of H)

I) vs D) - on this one we can probably remove I) - the reason I added it was because of the way I specified the context to the test tool - so its probably more of a test of the testing infrastructure than Docker itself.

Net: remove J2) and I)   ok?
But overall, I'm not sure we should be too concerned with dup testcases - its holes in the coverage that a bigger concern.  One or two extra ones isn't a big deal to me.

duglin commented:
@tiborvass you ok with removing J2 and i ?

ahmetalpbalkan commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -4526,85 +4527,82 @@ func TestBuildRenamedDockerfile(t *testing.T) {
+		t.Fatalf("test9 should have used root Dockerfile, output:%s", out)
+	}
+
+	out, _, err = dockerCmdInDir(t, ctx.Dir+"/files", "build", "-f", "dFile2", "-t", "test10", ".")
```
not gonna work on windows

duglin replied:
fixed

ahmetalpbalkan commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -4526,85 +4527,82 @@ func TestBuildRenamedDockerfile(t *testing.T) {
 
-	out, _, err = dockerCmdInDir(t, ctx.Dir+"/files", "build", "-f", ctx.Dir+"/files/Dockerfile", "-t", "test6", "..")
-
+	out, _, err = dockerCmdInDir(t, ctx.Dir+"/files", "build", "-f", ctx.Dir+"/files/Dockerfile", "-t", "test7", "..")
```
use filepath, not gonna work on windows

duglin replied:
fixed
